### ISSUE TITLE
Different behaviour depending on parameter names in model

### ISSUE THREAD

**Issue by wuerfelfreak** [PhiLFitters/kafe2#241]
**Describe the bug**
Creating a kafe2.xy_fit fails when the model is
```
def modell(xwfa, a, b, c):
    return 1
```
resulting in `TypeError: modell() missing 1 required positional argument: 'c'` but runs flawlessly when the model is
```
def modell(xwff, a, b, c):
    return 1
```
Theo only difference is in the argument `xwfa` vs. `xwff`
There are many other examples where it either fails (`omega,xwfa`) or works as expected (`x,y,xwff`)

**To Reproduce**
Install kafe2==2.8.0

Run
```
import kafe2

x_list = [1,2]
y_list = [1,2]

def modell(xwfa, a, b, c):
    return 1

fit_P_0 = kafe2.xy_fit(modell, x_list, y_list, x_error=1,y_error=1)
```

or 

```
import kafe2

x_list = [1,2]
y_list = [1,2]

def modell(xwff, a, b, c):
    return 1

fit_P_0 = kafe2.xy_fit(modell, x_list, y_list, x_error=1,y_error=1)
```

**Expected behavior**
The two example should behave exactly the same. There should not be an error with `xwfa`. 

**Traceback**
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_150131/424664083.py in <module>
      7     return 1
      8 
----> 9 fit_P_0 = kafe2.xy_fit(modell, x_list, y_list, x_error=1,y_error=1)

~/.local/lib/python3.10/site-packages/kafe2/fit/util/wrapper.py in xy_fit(model_function, x_data, y_data, p0, dp0, x_error, y_error, x_error_rel, y_error_rel, x_error_cor, y_error_cor, x_error_cor_rel, y_error_cor_rel, errors_rel_to_model, limits, fixed, constraints, report, profile, save)
    358         _fit = XYFit([x_data, y_data])
    359     else:
--> 360         _fit = XYFit([x_data, y_data], model_function)
    361 
    362     def _add_error_to_fit(axis, error, correlated=False, relative=False):

~/.local/lib/python3.10/site-packages/kafe2/fit/xy/fit.py in __init__(self, xy_data, model_function, cost_function, minimizer, minimizer_kwargs, dynamic_error_algorithm)
     66         :type minimizer_kwargs: dict
     67         """
---> 68         super(XYFit, self).__init__(
     69             data=xy_data, model_function=model_function, cost_function=cost_function,
     70             minimizer=minimizer, minimizer_kwargs=minimizer_kwargs,

~/.local/lib/python3.10/site-packages/kafe2/fit/_base/fit.py in __init__(self, data, model_function, cost_function, minimizer, minimizer_kwargs, dynamic_error_algorithm)
    124         # set the data after the cost_function has been set and nexus has been initialized
    125         if data is not None:
--> 126             self.data = data
    127 
    128     # -- private methods

~/.local/lib/python3.10/site-packages/kafe2/fit/_base/fit.py in data(self, new_data)
    460         if not _data_and_cost_compatible:
    461             raise ValueError('Fit data and cost function are not compatible: %s' % _reason)
--> 462         self._set_new_parametric_model()
    463         self._param_model._on_error_change_callbacks = [self._on_error_change]
    464 

~/.local/lib/python3.10/site-packages/kafe2/fit/xy/fit.py in _set_new_parametric_model(self)
    131 
    132     def _set_new_parametric_model(self):
--> 133         self._param_model = XYParametricModel(
    134             self.x_model,
    135             self._model_function,

~/.local/lib/python3.10/site-packages/kafe2/fit/xy/model.py in __init__(self, x_data, model_func, model_parameters)
     31         """
     32         _x_data_array = np.array(x_data)
---> 33         _y_data = model_func(_x_data_array, *model_parameters)
     34         if np.isscalar(_y_data):
     35             _y_data = np.ones_like(_x_data_array) * _y_data

~/.local/lib/python3.10/site-packages/kafe2/fit/_base/model.py in __call__(self, *args, **kwargs)
    193 
    194     def __call__(self, *args, **kwargs):
--> 195         return self._callable(*args, **kwargs)
    196 
    197     @property

TypeError: modell() missing 1 required positional argument: 'c'
```

**System:**
 - Windows11 sowie Ubuntu 22.04 getestet.
 - Python 3.10.12

**Comment by MitchiLaser** [comment]
I can confirm that this error also occurs with the newest release version `2.8.2`.

**Comment by MitchiLaser** [comment]
Here are some further investigations: The example fails only when the name of the first variable is `xwfa`, `xwfb`or `xwfc`. By adding another parameter `d` to the list of arguments it will also fail when the first parameter is set do `xwfd`, e.g. the following example is going to fail with the same stacktrace as the examples mentioned in the issue description: 
```
import kafe2

x_list = [1,2]
y_list = [1,2]

def modell(xwfd, a, b, c, d):
    return 1

fit_P_0 = kafe2.xy_fit(modell, x_list, y_list, x_error=1,y_error=1)
```
whereas this one is going to work:
```
import kafe2

x_list = [1,2]
y_list = [1,2]

def modell(xwfd, a, b, c):
    return 1

fit_P_0 = kafe2.xy_fit(modell, x_list, y_list, x_error=1,y_error=1)
```

After I took a further look: **Kafe2 only works when the name of a parameter is not contained at any position in another parameter**. I realized that with the following definition of the model function `def model1(test_name, a, test): return 1` it will fail with the message `TypeError: modell() missing 2 required positional arguments: 'a' and 'test'` whereas the definition `def model1(test_nme, a, test): return 1` only complains about a missing required argument for `'test'` and renaming the first argument to `tes_name'` by removing a letter 't' it stops complaining. This behavior also applies to similar examples.

**Comment by cverstege** [comment]
My first guess would be that there are some issues with this function:
https://github.com/PhiLFitters/kafe2/blob/3806d12d0d5340843d438c45cf142bc80d10d01b/kafe2/fit/_base/model.py#L132-L138

I'm quite busy right now, I could take a deeper look in 1-2 weeks.

**Comment by cverstege** [comment]
My second guess is something with these list conversion is not working as intended:
https://github.com/PhiLFitters/kafe2/blob/3806d12d0d5340843d438c45cf142bc80d10d01b/kafe2/fit/_base/model.py#L221
https://github.com/PhiLFitters/kafe2/blob/3806d12d0d5340843d438c45cf142bc80d10d01b/kafe2/fit/_base/model.py#L229

**Comment by JohannesGaessler** [comment]
The issue was caused by incorrect logic when constructing a dictionary with model function default values. If a parameter name is a substring of the independent variable name then that parameter was incorrectly skipped so the model function receives too few values. I pushed a fix to the dev branch, please confirm whether it works.

**Comment by wuerfelfreak** [comment]
I can confirm that the error I described doesn't occur in the new version.

**Comment by MitchiLaser** [comment]
I tried the example in the issue description and the ones where I encountered some problems: Your fix seems to work indeed. Also all the kafe2 examples in the examples directory are still running fine. Currently the linter is still complaining but after some small changes I will create a new release.

**Comment by MitchiLaser** [comment]
This problem is now fixed in the new release v2.8.3
